File: tests/orchestrator/test_hooks.py
"""Tests for SDK hooks in SuperClaude.Orchestrator.hooks."""

import pytest

from SuperClaude.Orchestrator.evidence import EvidenceCollector
from SuperClaude.Orchestrator.hooks import (
    DANGEROUS_PATTERNS,
    create_evidence_hooks,
    create_logging_hooks,
    create_safety_hooks,
    create_sdk_hooks,
    merge_hooks,
)


class TestDangerousPatterns:
    """Tests for dangerous command patterns."""

    def test_dangerous_patterns_exists(self):
        """Test that dangerous patterns list exists."""
        assert len(DANGEROUS_PATTERNS) > 0

    def test_includes_rm_rf_root(self):
        """Test includes rm -rf / pattern."""
        assert "rm -rf /" in DANGEROUS_PATTERNS

    def test_includes_force_push_main(self):
        """Test includes force push to main."""
        assert "git push --force origin main" in DANGEROUS_PATTERNS

    def test_includes_drop_table(self):
        """Test includes DROP TABLE pattern."""
        assert "DROP TABLE" in DANGEROUS_PATTERNS


class TestCreateSafetyHooks:
    """Tests for create_safety_hooks function."""

    def test_returns_hook_config(self):
        """Test returns hook configuration dict."""
        hooks = create_safety_hooks()

        assert isinstance(hooks, dict)
        assert "PreToolUse" in hooks

    def test_pretooluse_has_matchers(self):
        """Test PreToolUse hooks have matchers."""
        hooks = create_safety_hooks()

        assert len(hooks["PreToolUse"]) > 0
        assert "matcher" in hooks["PreToolUse"][0]


class TestBlockDangerousCommands:
    """Tests for blocking dangerous bash commands."""

    @pytest.fixture
    def safety_hooks(self):
        """Create safety hooks."""
        return create_safety_hooks()

    @pytest.mark.asyncio
    async def test_blocks_rm_rf_root(self, safety_hooks):
        """Test blocks rm -rf / command."""
        hook_fn = safety_hooks["PreToolUse"][0]["hooks"][0]

        result = await hook_fn(
            {
                "hook_event_name": "PreToolUse",
                "tool_name": "Bash",
                "tool_input": {"command": "rm -rf /"},
            },
            None,
            {},
        )

        assert "hookSpecificOutput" in result
        assert result["hookSpecificOutput"]["permissionDecision"] == "deny"

    @pytest.mark.asyncio
    async def test_blocks_git_reset_hard(self, safety_hooks):
        """Test blocks git reset --hard."""
        hook_fn = safety_hooks["PreToolUse"][0]["hooks"][0]

        result = await hook_fn(
            {
                "hook_event_name": "PreToolUse",
                "tool_name": "Bash",
                "tool_input": {"command": "git reset --hard HEAD~5"},
            },
            None,
            {},
        )

        assert result.get("hookSpecificOutput", {}).get("permissionDecision") == "deny"

    @pytest.mark.asyncio
    async def test_allows_safe_commands(self, safety_hooks):
        """Test allows safe commands."""
        hook_fn = safety_hooks["PreToolUse"][0]["hooks"][0]

        result = await hook_fn(
            {
                "hook_event_name": "PreToolUse",
                "tool_name": "Bash",
                "tool_input": {"command": "ls -la"},
            },
            None,
            {},
        )

        # Empty result means allow
        assert result == {}

    @pytest.mark.asyncio
    async def test_ignores_non_bash_tools(self, safety_hooks):
        """Test ignores non-Bash tools."""
        hook_fn = safety_hooks["PreToolUse"][0]["hooks"][0]

        result = await hook_fn(
            {
                "hook_event_name": "PreToolUse",
                "tool_name": "Read",
                "tool_input": {"file_path": "test.py"},
            },
            None,
            {},
        )

        assert result == {}


class TestValidateFilePaths:
    """Tests for file path validation hooks."""

    @pytest.fixture
    def safety_hooks(self):
        """Create safety hooks."""
        return create_safety_hooks()

    @pytest.mark.asyncio
    async def test_blocks_etc_writes(self, safety_hooks):
        """Test blocks writes to /etc/."""
        hook_fn = safety_hooks["PreToolUse"][1]["hooks"][0]

        result = await hook_fn(
            {
                "hook_event_name": "PreToolUse",
                "tool_name": "Write",
                "tool_input": {"file_path": "/etc/passwd"},
            },
            None,
            {},
        )

        assert result.get("hookSpecificOutput", {}).get("permissionDecision") == "deny"

    @pytest.mark.asyncio
    async def test_blocks_ssh_writes(self, safety_hooks):
        """Test blocks writes to ~/.ssh/."""
        hook_fn = safety_hooks["PreToolUse"][1]["hooks"][0]

        result = await hook_fn(
            {
                "hook_event_name": "PreToolUse",
                "tool_name": "Write",
                "tool_input": {"file_path": "~/.ssh/id_rsa"},
            },
            None,
            {},
        )

        assert result.get("hookSpecificOutput", {}).get("permissionDecision") == "deny"

    @pytest.mark.asyncio
    async def test_blocks_env_files(self, safety_hooks):
        """Test blocks writes to .env files."""
        hook_fn = safety_hooks["PreToolUse"][1]["hooks"][0]

        result = await hook_fn(
            {
                "hook_event_name": "PreToolUse",
                "tool_name": "Write",
                "tool_input": {"file_path": "/project/.env"},
            },
            None,
            {},
        )

        assert result.get("hookSpecificOutput", {}).get("permissionDecision") == "deny"

    @pytest.mark.asyncio
    async def test_allows_safe_paths(self, safety_hooks):
        """Test allows safe file paths."""
        hook_fn = safety_hooks["PreToolUse"][1]["hooks"][0]

        result = await hook_fn(
            {
                "hook_event_name": "PreToolUse",
                "tool_name": "Write",
                "tool_input": {"file_path": "/home/user/project/src/main.py"},
            },
            None,
            {},
        )

        assert result == {}


class TestCreateEvidenceHooks:
    """Tests for create_evidence_hooks function."""

    @pytest.fixture
    def evidence(self):
        """Create evidence collector."""
        return EvidenceCollector()

    def test_returns_hook_config(self, evidence):
        """Test returns hook configuration dict."""
        hooks = create_evidence_hooks(evidence)

        assert isinstance(hooks, dict)
        assert "PostToolUse" in hooks
        assert "Stop" in hooks
        assert "SubagentStop" in hooks

    def test_posttooluse_has_multiple_matchers(self, evidence):
        """Test PostToolUse has multiple matchers."""
        hooks = create_evidence_hooks(evidence)

        assert len(hooks["PostToolUse"]) >= 2


class TestCollectFileChanges:
    """Tests for file change collection hooks."""

    @pytest.fixture
    def evidence(self):
        """Create evidence collector."""
        return EvidenceCollector()

    @pytest.fixture
    def evidence_hooks(self, evidence):
        """Create evidence hooks."""
        return create_evidence_hooks(evidence)

    @pytest.mark.asyncio
    async def test_tracks_writes(self, evidence, evidence_hooks):
        """Test tracks file writes."""
        hook_fn = evidence_hooks["PostToolUse"][0]["hooks"][0]

        await hook_fn(
            {
                "hook_event_name": "PostToolUse",
                "tool_name": "Write",
                "tool_input": {"file_path": "new.py", "content": "code\nmore code"},
            },
            None,
            {},
        )

        assert "new.py" in evidence.files_written

    @pytest.mark.asyncio
    async def test_tracks_edits(self, evidence, evidence_hooks):
        """Test tracks file edits."""
        hook_fn = evidence_hooks["PostToolUse"][0]["hooks"][0]

        await hook_fn(
            {
                "hook_event_name": "PostToolUse",
                "tool_name": "Edit",
                "tool_input": {
                    "file_path": "existing.py",
                    "old_string": "old",
                    "new_string": "new",
                },
            },
            None,
            {},
        )

        assert "existing.py" in evidence.files_edited

    @pytest.mark.asyncio
    async def test_tracks_reads(self, evidence, evidence_hooks):
        """Test tracks file reads."""
        hook_fn = evidence_hooks["PostToolUse"][0]["hooks"][0]

        await hook_fn(
            {
                "hook_event_name": "PostToolUse",
                "tool_name": "Read",
                "tool_input": {"file_path": "readme.md"},
            },
            None,
            {},
        )

        assert "readme.md" in evidence.files_read


class TestCollectCommandResults:
    """Tests for command result collection hooks."""

    @pytest.fixture
    def evidence(self):
        """Create evidence collector."""
        return EvidenceCollector()

    @pytest.fixture
    def evidence_hooks(self, evidence):
        """Create evidence hooks."""
        return create_evidence_hooks(evidence)

    @pytest.mark.asyncio
    async def test_tracks_commands(self, evidence, evidence_hooks):
        """Test tracks bash commands."""
        hook_fn = evidence_hooks["PostToolUse"][1]["hooks"][0]

        await hook_fn(
            {
                "hook_event_name": "PostToolUse",
                "tool_name": "Bash",
                "tool_input": {"command": "ls -la"},
                "tool_response": "file1.py\nfile2.py",
            },
            None,
            {},
        )

        assert len(evidence.commands_run) == 1
        assert evidence.commands_run[0].command == "ls -la"

    @pytest.mark.asyncio
    async def test_parses_test_output(self, evidence, evidence_hooks):
        """Test parses test output from commands."""
        hook_fn = evidence_hooks["PostToolUse"][1]["hooks"][0]

        await hook_fn(
            {
                "hook_event_name": "PostToolUse",
                "tool_name": "Bash",
                "tool_input": {"command": "pytest tests/"},
                "tool_response": "===== 10 passed in 2.5s =====",
            },
            None,
            {},
        )

        assert evidence.tests_run is True
        assert len(evidence.test_results) == 1


class TestTrackAllTools:
    """Tests for tracking all tool invocations."""

    @pytest.fixture
    def evidence(self):
        """Create evidence collector."""
        return EvidenceCollector()

    @pytest.fixture
    def evidence_hooks(self, evidence):
        """Create evidence hooks."""
        return create_evidence_hooks(evidence)

    @pytest.mark.asyncio
    async def test_tracks_all_tools(self, evidence, evidence_hooks):
        """Test tracks all tool invocations."""
        # Get the hook without matcher (tracks all)
        hook_fn = evidence_hooks["PostToolUse"][2]["hooks"][0]

        await hook_fn(
            {
                "hook_event_name": "PostToolUse",
                "tool_name": "Glob",
                "tool_input": {"pattern": "*.py"},
                "tool_response": ["a.py", "b.py"],
            },
            "tool-use-123",
            {},
        )

        assert len(evidence.tool_invocations) == 1
        assert evidence.tool_invocations[0]["tool_name"] == "Glob"


class TestSessionHooks:
    """Tests for session lifecycle hooks."""

    @pytest.fixture
    def evidence(self):
        """Create evidence collector."""
        return EvidenceCollector()

    @pytest.fixture
    def evidence_hooks(self, evidence):
        """Create evidence hooks."""
        return create_evidence_hooks(evidence)

    @pytest.mark.asyncio
    async def test_on_stop(self, evidence, evidence_hooks):
        """Test Stop hook sets end time."""
        hook_fn = evidence_hooks["Stop"][0]["hooks"][0]

        await hook_fn(
            {
                "hook_event_name": "Stop",
                "session_id": "session-123",
            },
            None,
            {},
        )

        assert evidence.end_time is not None
        assert evidence.session_id == "session-123"

    @pytest.mark.asyncio
    async def test_on_subagent_stop(self, evidence, evidence_hooks):
        """Test SubagentStop hook tracks subagents."""
        hook_fn = evidence_hooks["SubagentStop"][0]["hooks"][0]

        await hook_fn(
            {
                "hook_event_name": "SubagentStop",
                "stop_hook_active": True,
            },
            "subagent-tool-use-456",
            {},
        )

        assert evidence.subagents_spawned == 1
        assert len(evidence.subagent_results) == 1


class TestCreateSdkHooks:
    """Tests for create_sdk_hooks function."""

    def test_combines_safety_and_evidence(self):
        """Test combines safety and evidence hooks."""
        evidence = EvidenceCollector()
        hooks = create_sdk_hooks(evidence)

        assert "PreToolUse" in hooks
        assert "PostToolUse" in hooks
        assert "Stop" in hooks
        assert "SubagentStop" in hooks

    def test_pretooluse_has_safety_hooks(self):
        """Test PreToolUse includes safety hooks."""
        evidence = EvidenceCollector()
        hooks = create_sdk_hooks(evidence)

        # Should have multiple PreToolUse matchers
        assert len(hooks["PreToolUse"]) >= 2


class TestCreateLoggingHooks:
    """Tests for create_logging_hooks function."""

    def test_returns_hook_config(self):
        """Test returns hook configuration."""
        logs = []
        hooks = create_logging_hooks(logs.append)

        assert "PreToolUse" in hooks
        assert "PostToolUse" in hooks

    @pytest.mark.asyncio
    async def test_logs_pre_tool_use(self):
        """Test logs PreToolUse events."""
        logs = []
        hooks = create_logging_hooks(logs.append)
        hook_fn = hooks["PreToolUse"][0]["hooks"][0]

        await hook_fn(
            {
                "hook_event_name": "PreToolUse",
                "tool_name": "Read",
                "tool_input": {"file_path": "test.py"},
            },
            None,
            {},
        )

        assert len(logs) == 1
        assert "[PRE]" in logs[0]
        assert "Read" in logs[0]

    @pytest.mark.asyncio
    async def test_logs_post_tool_use(self):
        """Test logs PostToolUse events."""
        logs = []
        hooks = create_logging_hooks(logs.append)
        hook_fn = hooks["PostToolUse"][0]["hooks"][0]

        await hook_fn(
            {
                "hook_event_name": "PostToolUse",
                "tool_name": "Read",
                "tool_response": "file contents",
            },
            None,
            {},
        )

        assert len(logs) == 1
        assert "[POST]" in logs[0]


class TestMergeHooks:
    """Tests for merge_hooks function."""

    def test_merges_pretooluse(self):
        """Test merges PreToolUse hooks."""
        config1 = {"PreToolUse": [{"matcher": "Bash", "hooks": ["hook1"]}]}
        config2 = {"PreToolUse": [{"matcher": "Write", "hooks": ["hook2"]}]}

        merged = merge_hooks(config1, config2)

        assert len(merged["PreToolUse"]) == 2

    def test_merges_different_events(self):
        """Test merges different event types."""
        config1 = {"PreToolUse": [{"hooks": ["hook1"]}]}
        config2 = {"PostToolUse": [{"hooks": ["hook2"]}]}

        merged = merge_hooks(config1, config2)

        assert "PreToolUse" in merged
        assert "PostToolUse" in merged

    def test_preserves_order(self):
        """Test preserves hook order."""
        config1 = {"PreToolUse": [{"hooks": ["first"]}]}
        config2 = {"PreToolUse": [{"hooks": ["second"]}]}

        merged = merge_hooks(config1, config2)

        assert merged["PreToolUse"][0]["hooks"] == ["first"]
        assert merged["PreToolUse"][1]["hooks"] == ["second"]

    def test_handles_empty_configs(self):
        """Test handles empty configurations."""
        config1 = {}
        config2 = {"PreToolUse": [{"hooks": ["hook"]}]}

        merged = merge_hooks(config1, config2)

        assert "PreToolUse" in merged

File: crates/superclaude-runtime/tests/test_hooks.rs
//! Integration tests for hooks module

use serde_json::{json, Value};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use superclaude_runtime::evidence::EvidenceCollector;
use superclaude_runtime::hooks::{
    create_evidence_hooks, create_logging_hooks, create_safety_hooks, create_sdk_hooks, merge_hooks,
    HookInput, HookOutput,
};

/// Helper to create a HookInput for testing
fn make_hook_input(
    event_name: &str,
    tool_name: &str,
    tool_input: HashMap<String, Value>,
) -> HookInput {
    HookInput {
        hook_event_name: event_name.to_string(),
        tool_name: tool_name.to_string(),
        tool_input,
        tool_response: Value::Null,
        session_id: "test-session".to_string(),
        stop_hook_active: false,
    }
}

#[tokio::test]
async fn test_safety_hooks_block_rm_rf_root() {
    let config = create_safety_hooks();

    // Find Bash matcher
    let bash_matcher = config
        .pre_tool_use
        .iter()
        .find(|m| m.matcher.as_deref() == Some("Bash"))
        .expect("Should have Bash matcher");

    let mut tool_input = HashMap::new();
    tool_input.insert("command".to_string(), json!("rm -rf /"));

    let input = make_hook_input("PreToolUse", "Bash", tool_input);
    let result = bash_matcher.hooks[0](input, None, HashMap::new()).await;

    assert!(result.hook_specific_output.is_some());
    let output = result.hook_specific_output.unwrap();
    assert_eq!(output.permission_decision, Some("deny".to_string()));
    assert!(output
        .permission_decision_reason
        .unwrap()
        .contains("Dangerous command"));
}

#[tokio::test]
async fn test_safety_hooks_block_git_reset_hard() {
    let config = create_safety_hooks();

    let bash_matcher = config
        .pre_tool_use
        .iter()
        .find(|m| m.matcher.as_deref() == Some("Bash"))
        .unwrap();

    let mut tool_input = HashMap::new();
    tool_input.insert("command".to_string(), json!("git reset --hard"));

    let input = make_hook_input("PreToolUse", "Bash", tool_input);
    let result = bash_matcher.hooks[0](input, None, HashMap::new()).await;

    assert!(result.hook_specific_output.is_some());
}

#[tokio::test]
async fn test_safety_hooks_allow_safe_commands() {
    let config = create_safety_hooks();

    let bash_matcher = config
        .pre_tool_use
        .iter()
        .find(|m| m.matcher.as_deref() == Some("Bash"))
        .unwrap();

    let safe_commands = vec!["ls -la", "git status", "npm install", "cargo build"];

    for command in safe_commands {
        let mut tool_input = HashMap::new();
        tool_input.insert("command".to_string(), json!(command));

        let input = make_hook_input("PreToolUse", "Bash", tool_input);
        let result = bash_matcher.hooks[0](input, None, HashMap::new()).await;

        assert!(
            result.hook_specific_output.is_none(),
            "Command '{}' should be allowed",
            command
        );
    }
}

#[tokio::test]
async fn test_safety_hooks_block_system_paths() {
    let config = create_safety_hooks();

    let write_matcher = config
        .pre_tool_use
        .iter()
        .find(|m| m.matcher.as_deref() == Some("Write|Edit"))
        .unwrap();

    let mut tool_input = HashMap::new();
    tool_input.insert("file_path".to_string(), json!("/etc/passwd"));

    let input = make_hook_input("PreToolUse", "Write", tool_input);
    let result = write_matcher.hooks[0](input, None, HashMap::new()).await;

    assert!(result.hook_specific_output.is_some());
    let output = result.hook_specific_output.unwrap();
    assert_eq!(output.permission_decision, Some("deny".to_string()));
}

#[tokio::test]
async fn test_safety_hooks_allow_user_paths() {
    let config = create_safety_hooks();

    let write_matcher = config
        .pre_tool_use
        .iter()
        .find(|m| m.matcher.as_deref() == Some("Write|Edit"))
        .unwrap();

    let mut tool_input = HashMap::new();
    tool_input.insert("file_path".to_string(), json!("/home/user/test.py"));

    let input = make_hook_input("PreToolUse", "Write", tool_input);
    let result = write_matcher.hooks[0](input, None, HashMap::new()).await;

    assert!(result.hook_specific_output.is_none());
}

#[tokio::test]
async fn test_evidence_hooks_record_file_write() {
    let evidence = Arc::new(Mutex::new(EvidenceCollector::new()));
    let config = create_evidence_hooks(Arc::clone(&evidence));

    let write_matcher = config
        .post_tool_use
        .iter()
        .find(|m| m.matcher.as_deref() == Some("Write|Edit|Read"))
        .unwrap();

    let mut tool_input = HashMap::new();
    tool_input.insert("file_path".to_string(), json!("test.py"));
    tool_input.insert("content".to_string(), json!("line1\nline2\nline3"));

    let input = make_hook_input("PostToolUse", "Write", tool_input);
    write_matcher.hooks[0](input, None, HashMap::new()).await;

    let ev = evidence.lock().unwrap();
    assert_eq!(ev.files_written.len(), 1);
    assert_eq!(ev.files_written[0], "test.py");
    assert_eq!(ev.file_changes.len(), 1);
    assert_eq!(ev.file_changes[0].action, "write");
    assert_eq!(ev.file_changes[0].lines_changed, 3);
}

#[tokio::test]
async fn test_evidence_hooks_record_file_edit() {
    let evidence = Arc::new(Mutex::new(EvidenceCollector::new()));
    let config = create_evidence_hooks(Arc::clone(&evidence));

    let edit_matcher = config
        .post_tool_use
        .iter()
        .find(|m| m.matcher.as_deref() == Some("Write|Edit|Read"))
        .unwrap();

    let mut tool_input = HashMap::new();
    tool_input.insert("file_path".to_string(), json!("test.py"));
    tool_input.insert("old_string".to_string(), json!("old\ncode"));
    tool_input.insert("new_string".to_string(), json!("new\ncode\nmore"));

    let input = make_hook_input("PostToolUse", "Edit", tool_input);
    edit_matcher.hooks[0](input, None, HashMap::new()).await;

    let ev = evidence.lock().unwrap();
    assert_eq!(ev.files_edited.len(), 1);
    assert_eq!(ev.files_edited[0], "test.py");
    assert_eq!(ev.file_changes[0].action, "edit");
    // abs(2 - 1) + 1 = 2 lines changed
    assert_eq!(ev.file_changes[0].lines_changed, 2);
}

#[tokio::test]
async fn test_evidence_hooks_record_file_read() {
    let evidence = Arc::new(Mutex::new(EvidenceCollector::new()));
    let config = create_evidence_hooks(Arc::clone(&evidence));

    let read_matcher = config
        .post_tool_use
        .iter()
        .find(|m| m.matcher.as_deref() == Some("Write|Edit|Read"))
        .unwrap();

    let mut tool_input = HashMap::new();
    tool_input.insert("file_path".to_string(), json!("test.py"));

    let input = make_hook_input("PostToolUse", "Read", tool_input);
    read_matcher.hooks[0](input, None, HashMap::new()).await;

    let ev = evidence.lock().unwrap();
    assert_eq!(ev.files_read.len(), 1);
    assert_eq!(ev.files_read[0], "test.py");
}

#[tokio::test]
async fn test_evidence_hooks_record_command() {
    let evidence = Arc::new(Mutex::new(EvidenceCollector::new()));
    let config = create_evidence_hooks(Arc::clone(&evidence));

    let bash_matcher = config
        .post_tool_use
        .iter()
        .find(|m| m.matcher.as_deref() == Some("Bash"))
        .unwrap();

    let mut tool_input = HashMap::new();
    tool_input.insert("command".to_string(), json!("ls -la"));

    let mut input = make_hook_input("PostToolUse", "Bash", tool_input);
    input.tool_response = json!("file1\nfile2\nfile3");

    bash_matcher.hooks[0](input, None, HashMap::new()).await;

    let ev = evidence.lock().unwrap();
    assert_eq!(ev.commands_run.len(), 1);
    assert_eq!(ev.commands_run[0].command, "ls -la");
}

#[tokio::test]
async fn test_evidence_hooks_track_all_tools() {
    let evidence = Arc::new(Mutex::new(EvidenceCollector::new()));
    let config = create_evidence_hooks(Arc::clone(&evidence));

    // Find the hook with no matcher (tracks all tools)
    let all_tools_matcher = config
        .post_tool_use
        .iter()
        .find(|m| m.matcher.is_none())
        .unwrap();

    let mut tool_input = HashMap::new();
    tool_input.insert("test_key".to_string(), json!("test_value"));

    let input = make_hook_input("PostToolUse", "SomeTool", tool_input);
    all_tools_matcher.hooks[0](input, None, HashMap::new()).await;

    let ev = evidence.lock().unwrap();
    assert_eq!(ev.tool_invocations.len(), 1);
    assert_eq!(ev.tool_invocations[0].tool_name, "SomeTool");
}

#[tokio::test]
async fn test_evidence_hooks_stop() {
    let evidence = Arc::new(Mutex::new(EvidenceCollector::new()));
    let config = create_evidence_hooks(Arc::clone(&evidence));

    assert_eq!(config.stop.len(), 1);

    let input = make_hook_input("Stop", "", HashMap::new());
    config.stop[0](input, None, HashMap::new()).await;

    let ev = evidence.lock().unwrap();
    assert!(ev.end_time.is_some());
    assert_eq!(ev.session_id, "test-session");
}

#[tokio::test]
async fn test_evidence_hooks_subagent_stop() {
    let evidence = Arc::new(Mutex::new(EvidenceCollector::new()));
    let config = create_evidence_hooks(Arc::clone(&evidence));

    assert_eq!(config.subagent_stop.len(), 1);

    let input = make_hook_input("SubagentStop", "", HashMap::new());
    config.subagent_stop[0](input, Some("tool-123".to_string()), HashMap::new()).await;

    let ev = evidence.lock().unwrap();
    assert_eq!(ev.subagents_spawned, 1);
    assert_eq!(ev.subagent_results.len(), 1);
}

#[tokio::test]
async fn test_create_sdk_hooks() {
    let evidence = Arc::new(Mutex::new(EvidenceCollector::new()));
    let config = create_sdk_hooks(evidence);

    // Should have safety + evidence hooks combined
    assert!(config.pre_tool_use.len() >= 2); // Safety hooks
    assert!(config.post_tool_use.len() >= 3); // Evidence hooks
    assert_eq!(config.stop.len(), 1);
    assert_eq!(config.subagent_stop.len(), 1);
}

#[tokio::test]
async fn test_merge_hooks() {
    let safety = create_safety_hooks();
    let evidence = Arc::new(Mutex::new(EvidenceCollector::new()));
    let evidence_hooks = create_evidence_hooks(evidence);

    let merged = merge_hooks(vec![safety, evidence_hooks]);

    assert!(merged.pre_tool_use.len() >= 2);
    assert!(merged.post_tool_use.len() >= 3);
    assert_eq!(merged.stop.len(), 1);
    assert_eq!(merged.subagent_stop.len(), 1);
}

#[tokio::test]
async fn test_logging_hooks() {
    let logs = Arc::new(Mutex::new(Vec::new()));
    let logs_clone = Arc::clone(&logs);

    let log_fn = move |msg: String| {
        logs_clone.lock().unwrap().push(msg);
    };

    let config = create_logging_hooks(log_fn);

    // Test PreToolUse logging
    let pre_matcher = config.pre_tool_use.iter().find(|m| m.matcher.is_none()).unwrap();

    let mut tool_input = HashMap::new();
    tool_input.insert("command".to_string(), json!("test command"));

    let input = make_hook_input("PreToolUse", "Bash", tool_input);
    pre_matcher.hooks[0](input, None, HashMap::new()).await;

    let logged = logs.lock().unwrap();
    assert_eq!(logged.len(), 1);
    assert!(logged[0].contains("[PRE]"));
    assert!(logged[0].contains("Bash"));
}

#[tokio::test]
async fn test_hook_ordering() {
    let evidence = Arc::new(Mutex::new(EvidenceCollector::new()));
    let config = create_sdk_hooks(Arc::clone(&evidence));

    // Safety hooks should be first in PreToolUse (block before evidence)
    assert!(config.pre_tool_use.len() >= 2);

    // Test that dangerous command is blocked BEFORE evidence collection
    let bash_matcher = config
        .pre_tool_use
        .iter()
        .find(|m| m.matcher.as_deref() == Some("Bash"))
        .unwrap();

    let mut tool_input = HashMap::new();
    tool_input.insert("command".to_string(), json!("rm -rf /"));

    let input = make_hook_input("PreToolUse", "Bash", tool_input);
    let result = bash_matcher.hooks[0](input, None, HashMap::new()).await;

    // Should be blocked
    assert!(result.hook_specific_output.is_some());

    // Evidence should not have recorded it (PreToolUse doesn't record)
    let ev = evidence.lock().unwrap();
    assert_eq!(ev.commands_run.len(), 0);
}

#[tokio::test]
async fn test_multiple_hooks_same_tool() {
    let evidence = Arc::new(Mutex::new(EvidenceCollector::new()));
    let config = create_sdk_hooks(evidence);

    // PostToolUse should have multiple hooks for file operations
    let file_matchers: Vec<_> = config
        .post_tool_use
        .iter()
        .filter(|m| {
            m.matcher
                .as_ref()
                .map(|s| s.contains("Write"))
                .unwrap_or(false)
        })
        .collect();

    assert!(file_matchers.len() >= 1);
}
